Compilation fails when using libtins-4.0.

```
  CXX      src/compactor-baseoutputwriter.o
In file included from ../src/dnsmessage.hpp:22:0,
                 from ../src/baseoutputwriter.hpp:20,
                 from ../src/baseoutputwriter.cpp:15:
../src/capturedns.hpp: In member function ‘virtual CaptureDNS* CaptureDNS::clone() const’:
../src/capturedns.hpp:914:36: error: invalid new-expression of abstract class type ‘CaptureDNS’
         return new CaptureDNS(*this);
                                    ^
../src/capturedns.hpp:74:7: note:   because the following virtual functions are pure within ‘CaptureDNS’:
 class CaptureDNS : public Tins::PDU
       ^~~~~~~~~~
In file included from /usr/local/include/tins/dns.h:39:0,
                 from /usr/local/include/tins/tins.h:33,
                 from ../src/ipaddress.hpp:22,
                 from ../src/configuration.hpp:21,
                 from ../src/baseoutputwriter.cpp:13:
/usr/local/include/tins/pdu.h:516:18: note:     virtual void Tins::PDU::write_serialization(uint8_t*, uint32_t)
     virtual void write_serialization(uint8_t* buffer, uint32_t total_sz) = 0;
                  ^~~~~~~~~~~~~~~~~~~
In file included from ../src/baseoutputwriter.hpp:20:0,
                 from ../src/baseoutputwriter.cpp:15:
../src/dnsmessage.hpp: At global scope:
../src/dnsmessage.hpp:174:16: error: cannot declare field ‘DNSMessage::dns’ to be of abstract type ‘CaptureDNS’
     CaptureDNS dns;
                ^~~
make: *** [Makefile:1361: src/compactor-baseoutputwriter.o] Error 1
```

--- a/configure.ac
+++ b/configure.ac
@@ -92,7 +92,10 @@ AS_IF([test "x$with_tcmalloc" != xno],
             fi])
         ])
 
-PKG_CHECK_MODULES(libtins, libtins >= 3.4)
+PKG_CHECK_MODULES(libtins,
+                  libtins >= 4.0,
+                  [AC_DEFINE([HAVE_LIBTINS4], [1], [Define to 1 if using libtins version 4.0 or later])],
+                  [PKG_CHECK_MODULES(libtins, libtins >= 3.4)])
 AC_SUBST(libtins_LIBS)
 
 # Checks for header files.

--- a/src/blockcborreader.cpp
+++ b/src/blockcborreader.cpp
@@ -11,6 +11,7 @@
  */
 
 #include <algorithm>
+#include <iomanip>
 #include <vector>
 
 #include "config.h"

--- a/src/capturedns.cpp
+++ b/src/capturedns.cpp
@@ -1033,7 +1033,11 @@ namespace {
 
 // This is used by libtins only.
 // cppcheck-suppress unusedFunction
-void CaptureDNS::write_serialization(uint8_t* buffer, uint32_t total_sz, const PDU*)
+#ifdef HAVE_LIBTINS4
+void CaptureDNS::write_serialization(uint8_t* buffer, uint32_t total_sz)
+#else
+void CaptureDNS::write_serialization(uint8_t* buffer, uint32_t total_sz, const PDU *)
+#endif
 {
     LabelCompressionInfo lci;
     OutputBufferStream stream(buffer, total_sz);

--- a/src/capturedns.hpp
+++ b/src/capturedns.hpp
@@ -25,6 +25,8 @@
 #include <tins/tins.h>
 #include <tins/memory_helpers.h>
 
+#include "config.h"
+
 #include "bytestring.hpp"
 
 /**
@@ -1224,14 +1226,18 @@ private:
     /**
      * \brief write serialised version of the packet.
      *
-     * This function is for `libtins` use when serialising the packet. Don't
-     * use it.
+     * This function is for `libtins` use when serialising the packet.
+     *  Don't use it.
      *
      * \param buffer    the packet data.
      * \param total_sz  the packet length.
-     * \param parent    the parent PDU.
+     * \param parent    the parent PDU (libtins < 4.0 only).
      */
+#ifdef HAVE_LIBTINS4
+    void write_serialization(uint8_t* buffer, uint32_t total_sz);
+#else
     void write_serialization(uint8_t* buffer, uint32_t total_sz, const PDU* parent);
+#endif
 
     /**
      * \brief the DNS packet header.

--- a/src/sniffers.cpp
+++ b/src/sniffers.cpp
@@ -1,5 +1,5 @@
 /*
- * Copyright 2016-2017 Internet Corporation for Assigned Names and Numbers.
+ * Copyright 2016-2018 Internet Corporation for Assigned Names and Numbers.
  *
  * This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
@@ -14,6 +14,11 @@
 
 #include <tins/loopback.h>
 #include <tins/pktap.h>
+
+#include "config.h"
+#ifdef HAVE_LIBTINS4
+#include <tins/detail/pdu_helpers.h>
+#endif
 
 #include "log.hpp"
 

--- a/tests/capturedns_test.cpp
+++ b/tests/capturedns_test.cpp
@@ -18,6 +18,8 @@
 #include "capturedns.hpp"
 #undef private
 
+#include "config.h"
+
 SCENARIO("Serialising DNS packets", "[dnspacket]")
 {
     GIVEN("A sample DNS message with one compression")
@@ -57,7 +59,11 @@ SCENARIO("Serialising DNS packets", "[dnspacket]")
             REQUIRE(pkt_size == EXPECTED.size());
 
             std::vector<uint8_t> buf(pkt_size);
+#ifdef HAVE_LIBTINS4
+            msg.write_serialization(buf.data(), pkt_size);
+#else
             msg.write_serialization(buf.data(), pkt_size, nullptr);
+#endif
             REQUIRE(buf == EXPECTED);
         }
     }
@@ -109,7 +115,11 @@ SCENARIO("Serialising DNS packets", "[dnspacket]")
             REQUIRE(pkt_size == EXPECTED.size());
 
             std::vector<uint8_t> buf(pkt_size);
+#ifdef HAVE_LIBTINS4
+            msg.write_serialization(buf.data(), pkt_size);
+#else
             msg.write_serialization(buf.data(), pkt_size, nullptr);
+#endif
             REQUIRE(buf == EXPECTED);
         }
     }
@@ -161,7 +171,11 @@ SCENARIO("Serialising DNS packets", "[dnspacket]")
             REQUIRE(pkt_size == EXPECTED.size());
 
             std::vector<uint8_t> buf(pkt_size);
+#ifdef HAVE_LIBTINS4
+            msg.write_serialization(buf.data(), pkt_size);
+#else
             msg.write_serialization(buf.data(), pkt_size, nullptr);
+#endif
             REQUIRE(buf == EXPECTED);
         }
     }

